sympy/physics/units/__init__.py

```diff
@@ -230,39 +230,53 @@ def find_unit(quantity, unit_system="SI"):
 
     >>> from sympy.physics import units as u
     >>> u.find_unit('charge')
-    ['C', 'coulomb', 'coulombs', 'planck_charge', 'elementary_charge']
+    ['C', 'charge', 'coulomb', 'coulombs', 'planck_charge', 'elementary_charge']
     >>> u.find_unit(u.charge)
     ['C', 'coulomb', 'coulombs', 'planck_charge', 'elementary_charge']
     >>> u.find_unit("ampere")
-    ['ampere', 'amperes']
+    ['A', 'ampere', 'amperes', 'planck_current']
     >>> u.find_unit('volt')
-    ['volt', 'volts', 'electronvolt', 'electronvolts', 'planck_voltage']
+    ['V', 'v', 'volt', 'volts', 'voltage', 'electronvolt', 'electronvolts', 'planck_voltage']
     >>> u.find_unit(u.inch**3)[:9]
     ['L', 'l', 'cL', 'cl', 'dL', 'dl', 'mL', 'ml', 'liter']
     """
     unit_system = UnitSystem.get_unit_system(unit_system)
 
     import sympy.physics.units as u
-    rv = []
+    rv = set()
+    IA = [(i, getattr(u, i)) for i in dir(u)]
+    IA = [(i, a) for i, a in IA if isinstance(a, (Quantity, Dimension))]
     if isinstance(quantity, str):
-        rv = [i for i in dir(u) if quantity in i and isinstance(getattr(u, i), Quantity)]
-        dim = getattr(u, quantity)
-        if isinstance(dim, Dimension):
-            rv.extend(find_unit(dim))
+        if quantity == '':
+            return []
+        dim = getattr(u, quantity, None)
+        if dim is None:
+            for i, a in IA:
+                if quantity in i:
+                    rv.add(i)
+                    rv.update(find_unit(i))
```

Comment:
I think that this is matching too much false positives like
```
>>> find_unit('a')
['A', 'C', 'D', 'F', 'G', 'H', 'J', 'K', 'L', 'N', 'R', 'S', 'T', 'V', 'W', 'g', 'h', 'l', 'm', 's', 't', 'v', 'Bq', 'Da', 'Gy', 'Pa', 'Wb', 'Z0', 'au', 'cL', 'cd', 'cl', 'cm', 'dL', 'dl', 'dm', 'e0', 'eV', 'ft', 'ha', 'kg', 'km', 'ly', 'mL', 'me', 'mg', 'mi', 'ml', 'mm', 'ms', 'nm', 'ns', 'pa', 'pm', 'ps', 'sr', 'u0', 'ug', 'um', 'us', 'wb', 'yd', 'amu', 'atm', 'bar', 'day', 'deg', 'gee', 'kPa', 'kat', 'mho', 'mil', 'mmu', 'mol', 'nmi', 'ohm', 'psi', 'rad', 'amus', 'area', 'bars', 'days', 'feet', 'foot', 'gees', 'gram', 'gray', 'hbar', 'hour', 'inch', 'mass', 'mhos', 'mile', 'mmHg', 'mmus', 'mole', 'ohms', 'torr', 'volt', 'watt', 'yard', 'year', 'farad', 'grams', 'henry', 'hours', 'joule', 'katal', 'liter', 'meter', 'miles', 'moles', 'pound', 'quart', 'tesla', 'tonne', 'volts', 'watts', 'weber', 'yards', 'years', 'action', 'amount', 'ampere', 'charge', 'dalton', 'degree', 'farads', 'henrys', 'inches', 'joules', 'kelvin', 'liters', 'meters', 'micron', 'minute', 'newton', 'pascal', 'planck', 'pounds', 'quarts', 'radian', 'second', 'stefan', 'teslas', 'webers', 'amperes', 'candela', 'coulomb', 'degrees', 'dioptre', 'hectare', 'kelvins', 'microns', 'minutes', 'newtons', 'pascals', 'percent', 'radians', 'seconds', 'siemens', 'voltage', 'avogadro', 'candelas', 'coulombs', 'kilogram', 'percents', 'permille', 'becquerel', 'boltzmann', 'deciliter', 'decimeter', 'impedance', 'kilograms', 'kilometer', 'lightyear', 'microgram', 'milligram', 'nanometer', 'picometer', 'steradian', 'atmosphere', 'centiliter', 'centimeter', 'deciliters', 'decimeters', 'inductance', 'kilometers', 'lightyears', 'metric_ton', 'micrograms', 'micrometer', 'milligrams', 'milliliter', 'millimeter', 'nanometers', 'nanosecond', 'picometers', 'picosecond', 'steradians', 'angular_mil', 'atmospheres', 'capacitance', 'centiliters', 'centimeters', 'common_year', 'conductance', 'julian_year', 'micrometers', 'microsecond', 'milliliters', 'millimeters', 'millisecond', 'nanoseconds', 'picoseconds', 'planck_area', 'planck_mass', 'planck_time', 'temperature', 'acceleration', 'angular_mils', 'common_years', 'electronvolt', 'julian_years', 'microseconds', 'milliseconds', 'planck_force', 'planck_power', 'draconic_year', 'electronvolts', 'gaussian_year', 'magnetic_flux', 'nautical_mile', 'optical_power', 'planck_charge', 'planck_energy', 'planck_length', 'planck_volume', 'sidereal_year', 'tropical_year', 'draconic_years', 'gaussian_years', 'nautical_miles', 'planck_current', 'planck_density', 'planck_voltage', 'sidereal_years', 'tropical_years', 'avogadro_number', 'full_moon_cycle', 'milli_mass_unit', 'planck_momentum', 'planck_pressure', 'anomalistic_year', 'atomic_mass_unit', 'coulomb_constant', 'faraday_constant', 'full_moon_cycles', 'magnetic_density', 'planck_impedance', 'planck_intensity', 'vacuum_impedance', 'anomalistic_years', 'astronomical_unit', 'avogadro_constant', 'electric_constant', 'elementary_charge', 'magnetic_constant', 'astronomical_units', 'boltzmann_constant', 'electron_rest_mass', 'josephson_constant', 'molar_gas_constant', 'planck_temperature', 'amount_of_substance', 'planck_acceleration', 'vacuum_permeability', 'vacuum_permittivity', 'atomic_mass_constant', 'magnetic_flux_density', 'planck_energy_density', 'von_klitzing_constant', 'gravitational_constant', 'electric_force_constant', 'planck_angular_frequency', 'stefan_boltzmann_constant', 'acceleration_due_to_gravity']
```

Reply:
Should I remove the searching by a string in find units?

Reply:
I think the match should be guarded for units with same dimensions

Reply:
Or be smart like only matching the prefix or suffix.

Reply:
``` python 
>>> from sympy.physics import units as u
>>> u.find_unit('charge')
['C', 'coulomb', 'coulombs', 'planck_charge', 'elementary_charge']
>>> u.find_unit('c')      
['c', 'cL', 'cd', 'cl', 'cm', 'charge', 'candela', 'coulomb', 'current', 'candelas', 'coulombs', 'centiliter', 'centimeter', 'capacitance', 'centiliters', 'centimeters', 'common_year', 'conductance', 'common_years', 'speed_of_light', 'coulomb_constant']
>>> u.find_unit('l')      
['L', 'l', 'cL', 'cl', 'dL', 'dl', 'lx', 'ly', 'mL', 'ml', 'lux', 'liter', 'quart', 'length', 'liters', 'quarts', 'deciliter', 'lightyear', 'centiliter', 'deciliters', 'lightyears', 'luminosity', 'milliliter', 'centiliters', 'milliliters', 'planck_volume', 'luminous_intensity']
>>> u.find_unit('volt')
['V', 'v', 'volt', 'volts', 'planck_voltage']
```
got these results after minor changes, like for `volts` it is showing  `['V', 'v', 'volt', 'volts', 'planck_voltage']` instead of `['V', 'v', 'volt', 'volts', 'voltage', 'electronvolt', 'electronvolts', 'planck_voltage']`
May be because electronvolt is not an unit of volt it is related to energy , same happens `u.find_unit('coulomb')` as it is not showing `'coulomb_constant'` bcoz it's unit is not coulomb

Reply:
The intent with string searching is that sometimes you know part of the name but maybe not the spelling, so the results returned helps you find what you are looking for. So if you use a single letter it is not surprising that you get many results. The results could be filtered if you want the letter at the beginning or end `[i for i in results if i[0] = letter]`.

Reply:
Okay, I get in the docs that this does not attempt to do more than substring search.

Reply:
`['V', 'v', 'volt', 'volts', 'voltage', 'electronvolt', 'electronvolts', 'planck_voltage']` I think `electronvolt` is not a part of `volt` it is unit of energy, so the correct output may be`['V', 'v', 'volt', 'volts', 'planck_voltage']`
what's ur thought about it?

Reply:
@smichr  so what to do now?

Reply:
Don't break the "text input finds subtext in units" behavior.